### PR TITLE
Do 3 part of SB practice

### PR DESCRIPTION
**Pull request by khorkovenko**

### CODE REVIEW

File: store/src/main/java/com/book/store/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+@RequiredArgsConstructor
+public class OrderController {
+    private final OrderService orderService;
+
+    @Operation(summary = "Place an order")
+    @PostMapping
+    public ResponseEntity<OrderDto> placeOrder(@Valid @RequestBody OrderDto orderDto) {
+        OrderDto createdOrder = orderService.placeOrder(orderDto);
+        return new ResponseEntity<>(createdOrder, HttpStatus.CREATED);
+    }
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/controller/OrderController.java]: use static method to create. 
```suggestion
        return ResponseEntity.created(...); // builder pattern here
    }
```

File: store/src/main/java/com/book/store/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Operation(summary = "Retrieve user's order history")
+    @GetMapping
+    public ResponseEntity<Page<OrderDto>> getUserOrderHistory(
+            Pageable pageable
+    ) {
+        Page<OrderDto> orderHistory = orderService
+                .getUserOrderHistory(pageable);
+        return new ResponseEntity<>(orderHistory, HttpStatus.OK);
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/controller/OrderController.java]: Prefer using static method for creation
```suggestion
        return ResponseEntity.ok(orderHistory);
```

File: store/src/main/java/com/book/store/model/Order.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private Status status;
+
+    @NotNull
+    private BigDecimal total;
+
+    @NotNull
+    private LocalDateTime orderDate;
+
+    @NotBlank
+    private String shippingAddress;
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/model/Order.java]: There are validation annotations, you need to add database constraints, e.g. Using @Column
```suggestion
    private User user;

    @Enumerated(EnumType.STRING)
    private Status status;

    private BigDecimal total;

    @NotNull
    private LocalDateTime orderDate;


    private String shippingAddress;
```

File: store/src/main/java/com/book/store/controller/OrderController.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+@RestController
+@RequestMapping("/api/orders")
+@RequiredArgsConstructor
+public class OrderController {
+    private final OrderService orderService;
+
+    @Operation(summary = "Place an order")
+    @PostMapping
+    public ResponseEntity<OrderDto> placeOrder(@Valid @RequestBody OrderDto orderDto) {
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/controller/OrderController.java:27]: Take OrderDto - return OrderDto : weird decision
You are using @Valid for OrderDto, but it has no rule for validation. 
```suggestion
    public ResponseEntity<OrderDto> placeOrder(@Valid @RequestBody OrderDto orderDto) {
```

File: store/src/main/java/com/book/store/dto/OrderDto.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.book.store.model.Status;
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Set;
+import lombok.Data;
+
+@Data
+public class OrderDto {
+    private Long id;
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/dto/OrderDto.java:19]: any validation rules, but uses @Valid to trigger them

File: store/src/main/java/com/book/store/model/OrderItem.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    @ManyToOne
+    @NotNull
+    private Book book;
+
+    @NotNull
+    private int quantity;
+
+    @NotNull
+    private BigDecimal price;
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/model/OrderItem.java]: Remove validation annotation, you need to add database constraints

File: store/src/main/java/com/book/store/model/Status.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    PROCESSING,
+
+    @Enumerated(EnumType.STRING)
+    SHIPPED,
+
+    @Enumerated(EnumType.STRING)
+    DELIVERED,
+
+    @Enumerated(EnumType.STRING)
+    CANCELED
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/model/Status.java]: Check what is @Enumerated and why we are using it

File: store/src/main/java/com/book/store/service/OrderService.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.book.store.dto.OrderDto;
+import com.book.store.model.Status;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+
+public interface OrderService {
+    OrderDto placeOrder(OrderDto orderDto);
+
+    Page<OrderDto> getUserOrderHistory(Pageable pageable);
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/service/OrderService.java]: You dont need to return page, think why
```suggestion
    List/Set<OrderDto> getUserOrderHistory(Pageable pageable);
```

File: store/src/main/java/com/book/store/service/impl/OrderItemServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+public class OrderItemServiceImpl implements OrderItemService {
+    private final OrderItemRepository orderItemRepository;
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
+    public Set<OrderItemDto> getAllOrderItems(Long orderId) {
+        Set<OrderItem> orderItems = orderItemRepository.findAllByOrderId(orderId);
+        return orderItems.stream()
+                .map(orderItemMapper::mapToOrderItemDto)
+                .collect(Collectors.toSet());
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/service/impl/OrderItemServiceImpl.java]: inline code
```suggestion
        Set<OrderItem> orderItems = orderItemRepository.findAllByOrderId(orderId);
        return orderItems.stream()
                .map(orderItemMapper::mapToOrderItemDto)
                .collect(Collectors.toSet());
```

File: store/src/main/java/com/book/store/service/impl/OrderItemServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    public OrderItemDto getOrderItem(Long orderId, Long itemId) {
+        OrderItem orderItem = orderItemRepository
+                .findByIdAndOrderId(itemId, orderId)
+                .orElseThrow(()
+                        -> new EntityNotFoundException("OrderItem not found with id: " + itemId));
+        return orderItemMapper.mapToOrderItemDto(orderItem);
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/service/impl/OrderItemServiceImpl.java]: inline
```suggestion
        OrderItem orderItem = orderItemRepository
                .findByIdAndOrderId(itemId, orderId)
                .orElseThrow(()
                        -> new EntityNotFoundException("OrderItem not found with id: " + itemId));
        return orderItemMapper.mapToOrderItemDto(orderItem);
```

File: store/src/main/java/com/book/store/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    public OrderDto placeOrder(OrderDto orderDto) {
+        Order order = orderMapper.mapToOrderEntity(orderDto);
+        Order savedOrder = orderRepository.save(order);
+        return orderMapper.mapToOrderDto(savedOrder);
+    }
+
+    @Override
+    public Page<OrderDto> getUserOrderHistory(Pageable pageable) {
+        Page<Order> userOrders = orderRepository.findAll(pageable);
+        return userOrders.map(orderMapper::mapToOrderDto);
```
Comment (sokimaaa) [inline store/src/main/java/com/book/store/service/impl/OrderServiceImpl.java]: inline
```suggestion
        Page<Order> userOrders = orderRepository.findAll(pageable);
        return userOrders.map(orderMapper::mapToOrderDto);
```

File: store/src/main/java/com/book/store/dto/OrderDto.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    private Long userId;
+
+    @Valid
+    @NotEmpty
+    private Set<OrderItemDto> orderItems;
+
+    @NotNull
+    private LocalDateTime orderDate;
+
+    @NotNull
```
Comment (Sanderfuck) [inline store/src/main/java/com/book/store/dto/OrderDto.java]:
```suggestion
    @NotNull
    @Min(0)
```

Comment (Sanderfuck) [reply store/src/main/java/com/book/store/dto/OrderDto.java]: Add more validation of fields

File: store/src/main/java/com/book/store/model/OrderItem.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    @Column(nullable = false)
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
+    @JoinColumn(nullable = false)
+    private Order order;
+
+    @ManyToOne
```
Comment (Sanderfuck) [inline store/src/main/java/com/book/store/model/OrderItem.java]: Is it Many to One?

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Review by Rommelua — CHANGES_REQUESTED** [review]
image

**Review by koval-kn — APPROVED** [review]

**Pull request merged**